### ISSUE TITLE
The Google Maps JavaScript API does not support this browser

### ISSUE THREAD

**Issue by Georgevanvelzen** [wkhtmltopdf/wkhtmltopdf#4185]
**wkhtmltopdf version(s) affected**: 12.5.0 (and older) 

**OS information**
Windows Server 2016, Windows 10

**Description**
Google has updated their browser support and detection on November 13th 2018. https://developers.google.com/maps/documentation/javascript/browsersupport

Running WKHTMLTOPDF against a page that uses the Google API will now fail.

**How to reproduce**
wkhtmltopdf.exe --username USER --password PASS --window-status ready_to_print --debug-javascript https://mypage c:\temp\output.pdf

Output with debug javascript enabled: The Google Maps JavaScript API does not support this browser

**Expected behavior**
Prior to November 13th code would run without a problem.

**Possible Solution**  
Enable browser impersonation.

**Comment by Georgevanvelzen** [comment]
I should add that the window-status was set after the Google Map completed plotting, and as the Google API now no longer works with WKHTMLTOPDF the script would never get to set window-status, so WKHTMLTOPDF would just hang on the server.

I have moved window-status above the Google Map API so the a PDF can be generated (without a map) and manually add a map drawn in Chrome with Adobe PDF editor afterwards.

This isn't ideal as the process is now not fully automated.

**Comment by ovider** [comment]
I am having the same issue starting November 13, when the new Google Maps API version was released (v.3.35). 

I can confirm that specifically loading the google maps with the version parameter to the previous version (e.g. https://maps.googleapis.com/maps/api/js?key=ABCD&v=3.34) works fine.

**Comment by varunHo** [comment]
I am having the same issue. Google recommends to test in early Feb. So I hope there will be some fixes to resolve this issue as I am using 0.12.4.

Check https://developers.google.com/maps/documentation/javascript/versions for details.
"We recommend, in early February, you test your application using v=3.35."

**Comment by adoming3** [comment]
I'm having this issue as well. As others have stated, downgrading to v.3.34 is a temporary fix.

**Comment by shiva416** [comment]
Hello everyone, I'm having the same from 13th Feb and downgrading to V 3.34 does not help me out here. Do we have any solution for this?

**Comment by ovider** [comment]
The solution we have eventually went with was to use static maps in PDF Reports. There is a limitation that you can get static maps only for a max 640x480px, but it is what it is.

**Comment by varunHo** [comment]
The only other solution is to use version 0.13.0. Though it is just an alpha release, you can use it to generate image of Map (created using Google Map JS API) and include it in the final PDF.

**Comment by shiva416** [comment]
Hi Varun,
Could you please provide the Github link or a way we can download version 0.13.0.

**Comment by varunHo** [comment]
> Hi Varun,
> Could you please provide the Github link or a way we can download version 0.13.0.

https://wkhtmltopdf.org/downloads.html
Check "Bleeding Edge" section and download suitable version for your flavor.

**Comment by shiva416** [comment]
> > Hi Varun,
> > Could you please provide the Github link or a way we can download version 0.13.0.
> 
> https://wkhtmltopdf.org/downloads.html
> Check "Bleeding Edge" section and download suitable version for your flavor.

I did try to use 0.13.0 with the sample.html attached, still, I could get only a blank image. Not sure if I'm missing something here. Please help..

[Sample.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2870483/Sample.zip)

**Comment by varunHo** [comment]
> > > Hi Varun,
> > > Could you please provide the Github link or a way we can download version 0.13.0.
> > 
> > 
> > https://wkhtmltopdf.org/downloads.html
> > Check "Bleeding Edge" section and download suitable version for your flavor.
> 
> I did try to use 0.13.0 with the sample.html attached, still, I could get only a blank image. Not sure if I'm missing something here. Please help..
> 
> [Sample.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2870483/Sample.zip)

Try using command line argument "--window-status whatever" and you need to set the window status through JS once you are sure that Map has finished loading. You can listen to "tilesloaded" and "idle" events if required. Ultimately you need to set window.status = 'whatever' from JS.

**Comment by shiva416** [comment]
> > > > Hi Varun,
> > > > Could you please provide the Github link or a way we can download version 0.13.0.
> > > 
> > > 
> > > https://wkhtmltopdf.org/downloads.html
> > > Check "Bleeding Edge" section and download suitable version for your flavor.
> > 
> > 
> > I did try to use 0.13.0 with the sample.html attached, still, I could get only a blank image. Not sure if I'm missing something here. Please help..
> > [Sample.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2870483/Sample.zip)
> 
> Try using command line argument "--window-status whatever" and you need to set the window status through JS once you are sure that Map has finished loading. You can listen to "tilesloaded" and "idle" events if required. Ultimately you need to set window.status = 'whatever' from JS.

That doesn't help, included command line argument and also I added Window.status - 'whatever' under tilesloaded event.  Once I added these and run the command, there is no response after Step 1 of the converting process. I did understand with this Map was never finished loading.

Later, I have downloaded QT Web browser and tried to access maps.google.com and it displayed a message from "Maps cannot be displayed, your browser needs to be updated". 

I have a question, is there anyway we can pass Useragent in command line argument to wkhtmltoimage.exe so the the request header so that the map is loaded.

QT web user agent:
Mozilla/5.0 (Windows; U; Windows NT based; en-US) AppleWebKit/534.34 (KHTML, like Gecko) QtWeb Internet Browser/3.8.5 http://www.QtWeb.net

**Comment by varunHo** [comment]
> > > > > Hi Varun,
> > > > > Could you please provide the Github link or a way we can download version 0.13.0.
> > > > 
> > > > 
> > > > https://wkhtmltopdf.org/downloads.html
> > > > Check "Bleeding Edge" section and download suitable version for your flavor.
> > > 
> > > 
> > > I did try to use 0.13.0 with the sample.html attached, still, I could get only a blank image. Not sure if I'm missing something here. Please help..
> > > [Sample.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2870483/Sample.zip)
> > 
> > 
> > Try using command line argument "--window-status whatever" and you need to set the window status through JS once you are sure that Map has finished loading. You can listen to "tilesloaded" and "idle" events if required. Ultimately you need to set window.status = 'whatever' from JS.
> 
> That doesn't help, included command line argument and also I added Window.status - 'whatever' under tilesloaded event. Once I added these and run the command, there is no response after Step 1 of the converting process. I did understand with this Map was never finished loading.
> 
> Later, I have downloaded QT Web browser and tried to access maps.google.com and it displayed a message from "Maps cannot be displayed, your browser needs to be updated".
> 
> I have a question, is there anyway we can pass Useragent in command line argument to wkhtmltoimage.exe in the request header so that the map is loaded.
> 
> QT web user agent:
> Mozilla/5.0 (Windows; U; Windows NT based; en-US) AppleWebKit/534.34 (KHTML, like Gecko) QtWeb Internet Browser/3.8.5 http://www.QtWeb.net

You are receiving error - "Maps cannot be displayed, your browser needs to be updated" that indicates you are still using older version. 0.13.0 uses Qt 5.4.2. You will not get error with this version.

I don't think you can pass UserAgent as command line argument. Refer - https://wkhtmltopdf.org/usage/wkhtmltopdf.txt for details. Same set of arguments are used for wkhtmltoimage.exe

**Comment by uzielgl** [comment]
hey people! I'm using wkhtml 0.13 versión, and is rendering fine the maps. But I'm having an issue with the --page-width and --page-height option. When I use this 2 options with 'px' (example: --page-height '3600px' --page-width '3000px') I have the next message and is not respect the size:
QString::arg: Argument missing: , 3000
QString::arg: Argument missing: , 3600

But if I use the in units, works in a stranger way (using for example --page-height '12in' --page-width '10in'), the PDF is created with the inches, but the page is cut in several pages.

I could fix it setting the in to the double values (--page-height '24in' --page-width '20in').

But why that behavior?? why that error of "argument missing".

Thanks a lot in advance!

**Comment by shiva416** [comment]
> > > > > > Hi Varun,
> > > > > > Could you please provide the Github link or a way we can download version 0.13.0.
> > > > > 
> > > > > 
> > > > > https://wkhtmltopdf.org/downloads.html
> > > > > Check "Bleeding Edge" section and download suitable version for your flavor.
> > > > 
> > > > 
> > > > I did try to use 0.13.0 with the sample.html attached, still, I could get only a blank image. Not sure if I'm missing something here. Please help..
> > > > [Sample.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2870483/Sample.zip)
> > > 
> > > 
> > > Try using command line argument "--window-status whatever" and you need to set the window status through JS once you are sure that Map has finished loading. You can listen to "tilesloaded" and "idle" events if required. Ultimately you need to set window.status = 'whatever' from JS.
> > 
> > 
> > That doesn't help, included command line argument and also I added Window.status - 'whatever' under tilesloaded event. Once I added these and run the command, there is no response after Step 1 of the converting process. I did understand with this Map was never finished loading.
> > Later, I have downloaded QT Web browser and tried to access maps.google.com and it displayed a message from "Maps cannot be displayed, your browser needs to be updated".
> > I have a question, is there anyway we can pass Useragent in command line argument to wkhtmltoimage.exe in the request header so that the map is loaded.
> > QT web user agent:
> > Mozilla/5.0 (Windows; U; Windows NT based; en-US) AppleWebKit/534.34 (KHTML, like Gecko) QtWeb Internet Browser/3.8.5 http://www.QtWeb.net
> 
> You are receiving error - "Maps cannot be displayed, your browser needs to be updated" that indicates you are still using older version. 0.13.0 uses Qt 5.4.2. You will not get error with this version.
> 
> I don't think you can pass UserAgent as command line argument. Refer - https://wkhtmltopdf.org/usage/wkhtmltopdf.txt for details. Same set of arguments are used for wkhtmltoimage.exe

Thanks, Varun.That really helped. 
I have been trying to use the stream in and stream out instead of working on physical files in c#, I was able to get this done with simpler HTML files but once I have larger HTML content nothing comes up. The same input HTML if I use in a physical file it works though. Have you ever come face this.

 Process p;
            StreamWriter stdin;
            ProcessStartInfo psi = new ProcessStartInfo();

            psi.FileName = Path.Combine(commandLocation, HtmlToPdfExePath);
            psi.WorkingDirectory = Path.GetDirectoryName(psi.FileName);

            // run the conversion utility
            psi.UseShellExecute = false;
            psi.CreateNoWindow = true;
            psi.RedirectStandardInput = true;
            psi.RedirectStandardOutput = true;
            psi.RedirectStandardError = true;

            // note: that we tell wkhtmltopdf to be quiet and not run scripts
            psi.Arguments = "-q -n --javascript-delay 2500  - -";

            p = Process.Start(psi);

            try
            {
                stdin = p.StandardInput;
                stdin.AutoFlush = true;
                stdin.Write(html);
                stdin.Dispose();
                p.WaitForExit();
                CopyStream(p.StandardOutput.BaseStream, pdf);
                p.StandardOutput.Close();
                pdf.Position = 0;

                

                return true;
            }
            catch
            {
                return false;

            }
            finally
            {
                p.Dispose();
            }

**Comment by leo-moxiworks** [comment]
After banging my head on this for a week, I was able to find this solution
```
      window.requestAnimationFrame = window.requestAnimationFrame || function () { return true; }
```

Google maps API is checking if your browser/window can use the requestAnimationFrame method and throwing an error if it cannot. You can override the method and just return true if your webpage does not use the requestAnimationFrame method.

```
 if (!_.y.requestAnimationFrame)
            throw _.U("controls").then(function(b) {
                b.pg(a)
            }),
            Error("The Google Maps JavaScript API does not support this browser.");
```

**Comment by shiva416** [comment]
> > > > > > Hi Varun,
> > > > > > Could you please provide the Github link or a way we can download version 0.13.0.
> > > > > 
> > > > > 
> > > > > https://wkhtmltopdf.org/downloads.html
> > > > > Check "Bleeding Edge" section and download suitable version for your flavor.
> > > > 
> > > > 
> > > > I did try to use 0.13.0 with the sample.html attached, still, I could get only a blank image. Not sure if I'm missing something here. Please help..
> > > > [Sample.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2870483/Sample.zip)
> > > 
> > > 
> > > Try using command line argument "--window-status whatever" and you need to set the window status through JS once you are sure that Map has finished loading. You can listen to "tilesloaded" and "idle" events if required. Ultimately you need to set window.status = 'whatever' from JS.
> > 
> > 
> > That doesn't help, included command line argument and also I added Window.status - 'whatever' under tilesloaded event. Once I added these and run the command, there is no response after Step 1 of the converting process. I did understand with this Map was never finished loading.
> > Later, I have downloaded QT Web browser and tried to access maps.google.com and it displayed a message from "Maps cannot be displayed, your browser needs to be updated".
> > I have a question, is there anyway we can pass Useragent in command line argument to wkhtmltoimage.exe in the request header so that the map is loaded.
> > QT web user agent:
> > Mozilla/5.0 (Windows; U; Windows NT based; en-US) AppleWebKit/534.34 (KHTML, like Gecko) QtWeb Internet Browser/3.8.5 http://www.QtWeb.net
> 
> You are receiving error - "Maps cannot be displayed, your browser needs to be updated" that indicates you are still using older version. 0.13.0 uses Qt 5.4.2. You will not get error with this version.
> 
> I don't think you can pass UserAgent as command line argument. Refer - https://wkhtmltopdf.org/usage/wkhtmltopdf.txt for details. Same set of arguments are used for wkhtmltoimage.exe
> 
> In case you are looking for a .NET component/Wrapper for wkhtmltopdf/wkhtmltoimage you can use NReco. https://www.nrecosite.com/pdf_generator_net.aspx

Thanks Varun..version 0.13.0 alpha resolved our issue. I cannot use the NRECO library as it would deploy only v 0.12.4 and there is no way to point 0.13.0 to it. Any way I could make work around the alpha version.

**Comment by exowanderer** [comment]
> only

From what I can tell, there is no 0.13 version for Mac OSX. Is that correct?

**Comment by exowanderer** [comment]
> After banging my head on this for a week, I was able to find this solution that works with wkhtmltopdf version 12.x. Add this code to your javascript to create a requestAnimationFrame function because the old version of QT web does not support this function.
> 
> ```
> window.requestAnimationFrame = window.requestAnimationFrame || function () { return true; }
> ```
> Google maps API is checking if your browser/window can use the requestAnimationFrame method and throwing an error if it cannot. You can override the method and just return true if your webpage does not use the requestAnimationFrame method. This works with google maps API version 3.35+.
> 
> ```
>  if (!_.y.requestAnimationFrame)
>             throw _.U("controls").then(function(b) {
>                 b.pg(a)
>             }),
>             Error("The Google Maps JavaScript API does not support this browser.");
> ```

Could you provide a html snippet to use with wkhtmltoimage or wkhtmltopdf that was able to generate a png/jpg/pdf with the Google Maps API after Nov 2018?

I am hoping to take your script and add my API_KEY to be able to test his functionality.

All of my efforts result in a *blank* pages.

------
For example, I am using `wkhtmltopdf` and `wkhtmltoimage` version 0.12.5 (stable) with the following html script (from this [Google Maps JavaScript API example](https://developers.google.com/maps/documentation/javascript/tutorial)):

```html
<!DOCTYPE html>
<html>
  <head>
    <meta name="viewport" content="initial-scale=1.0, user-scalable=no">
    <meta charset="utf-8">
    <title>Traffic Layer</title>
    <style>
      /* Always set the map height explicitly to define the size of the div
       * element that contains the map. */
      #map {
        height: 100%;
      }
      /* Optional: Makes the sample page fill the window. */
      html, body {
        height: 100%;
        margin: 0;
        padding: 0;
      }
    </style>
  </head>
  <body>
    <div id="map"></div>
    <script>
      function initMap() {
        var map = new google.maps.Map(document.getElementById('map'), {
          zoom: 12,
          center: {lat: -34.397, lng: 150.644}
        });

        var trafficLayer = new google.maps.TrafficLayer();
        trafficLayer.setMap(map);
        window.requestAnimationFrame = window.requestAnimationFrame || function () { return true; }
      }

    </script>
    <script async defer
    src="https://maps.googleapis.com/maps/api/js?key=API_KEY&callback=initMap&v=3.34">
    </script>
  </body>
</html>
```
[Note: I turned the v=3.34 on and off; I also placed the `window.request` section in and out of the `initMap` function.]

[Note: I do not often or easily program in JavaScript or HTML]

I saved that in a file named `google_maps_example.html`; and I used that as an input to `wkhtmltopdf` as such:

```
wkhtmltopdf --javascript-delay 5000  google_example_api.html google_example_api.pdf
```
or
```
wkhtmltoimage --javascript-delay 5000 --width 720 --height 1280  google_example_api.html google_example_api.png
```
or 
```
wkhtmltopdf --javascript-delay 10000 --page-height '24in' --page-width '20in'  google_example_api.html google_example_api.pdf
```
or many more combinations of javascript-delay, with or w/o page-height/width, with or w/o width/height in pixels.

I am on a Mac with 
```
Darwin Cloud.local 18.2.0 Darwin Kernel Version 18.2.0 x86_64
```

**Comment by leo-moxiworks** [comment]
> ```
>     <script>
>       function initMap() {
>         var map = new google.maps.Map(document.getElementById('map'), {
>           zoom: 12,
>           center: {lat: -34.397, lng: 150.644}
>         });
> 
>         var trafficLayer = new google.maps.TrafficLayer();
>         trafficLayer.setMap(map);
>         window.requestAnimationFrame = window.requestAnimationFrame || function () { return true; }
>       }
> 
>     </script>
> ```

Try moving the window.requestAnimationFrame line outside of your init function. You need to define this function before any of the google maps code runs.

```
    <script>
      window.requestAnimationFrame = window.requestAnimationFrame || function () { return true; }

      function initMap() {
        var map = new google.maps.Map(document.getElementById('map'), {
          zoom: 12,
          center: {lat: -34.397, lng: 150.644}
        });

        var trafficLayer = new google.maps.TrafficLayer();
        trafficLayer.setMap(map);
      }
    </script>
```

**Comment by exowanderer** [comment]
> 
> ```
>     <script>
>       window.requestAnimationFrame = window.requestAnimationFrame || function () { return true; }
> 
>       function initMap() {
>         var map = new google.maps.Map(document.getElementById('map'), {
>           zoom: 12,
>           center: {lat: -34.397, lng: 150.644}
>         });
> 
>         var trafficLayer = new google.maps.TrafficLayer();
>         trafficLayer.setMap(map);
>       }
>     </script>
> ```

Thank you for the suggestion. It did not work for me. It's not unlikely that I screwed something up. I've attached my exact code below in hopes that you will grab it, use it, and see if it produces a useful output for you.  That would at least explain that it's my fault or my local internet's fault. Either way, it may provide a path forward.

I put the below code into a file named "google_example_api.html" and tried

```
wkhtmltoimage --javascript-delay 10000 google_example_api.html google_example_api.jpg
```

Thank you for any help.

```html
<!DOCTYPE html>
<html>
  <head>
    <meta name="viewport" content="initial-scale=1.0, user-scalable=no">
    <meta charset="utf-8">
    <title>Traffic Layer</title>
    <style>
      /* Always set the map height explicitly to define the size of the div
       * element that contains the map. */
      #map {
        height: 100%;
      }
      /* Optional: Makes the sample page fill the window. */
      html, body {
        height: 100%;
        margin: 0;
        padding: 0;
      }
    </style>
  </head>
  <body>
    <div id="map"></div>
    <script>

      window.requestAnimationFrame = window.requestAnimationFrame || function () { return true; }

      function initMap() {
        var map = new google.maps.Map(document.getElementById('map'), {
          zoom: 12,
          center: {lat: 40.7128, lng: -74.0060}
        });

        var trafficLayer = new google.maps.TrafficLayer();
        trafficLayer.setMap(map);
        
      }

    </script>
    <script async defer
    src="https://maps.googleapis.com/maps/api/js?key=API_KEY&callback=initMap">
    </script>
  </body>
</html>
```

**Comment by leo-moxiworks** [comment]
@exowanderer try to add window.status = 'loaded' to the end of your JS execution

```
<script>

    window.requestAnimationFrame = window.requestAnimationFrame || function () { return true; }

    function initMap() {
        var map = new google.maps.Map(document.getElementById('map'), {
            zoom: 12,
            center: {lat: 40.7128, lng: -74.0060}
        });

        var trafficLayer = new google.maps.TrafficLayer();
        trafficLayer.setMap(map);

        window.status = 'loaded';

    }

</script>
```

Then use the window-status parameter in your wkhtmltoimage command
```
wkhtmltoimage --no-stop-slow-scripts --window-status 'loaded' google_example_api.html google_example_api.jpg
```

This will ensure all your JS runs before the PDF is generated

**Comment by exowanderer** [comment]
I sincerely appreciate the help. We might want to move this to a new issue.

I added the `window.status` line that you suggest; but it was not successful. I stored the html script below as before and ran the command that you suggested

`
wkhtmltoimage --no-stop-slow-scripts --window-status 'loaded' google_example_api.html google_example_api.jpg
`
`
wkhtmltopdf --no-stop-slow-scripts --window-status 'loaded' google_example_api.html google_example_api.pdf
`

but attained the same (negative) result: a 'jpg' with a single white bar (blank image). I also tried `wkhtmltopdf`, which resulted in a full page blank pdf

**In case it's relevant**, I am using OSX Mojave 10.14.3. Because I don't code in JS or HTML often (very python-centric), I don't know how to tell what version of either that I have. 

Could version issues be the problem?

Here is the script that I stored in `google_example_api.html`
```html
<!DOCTYPE html>
<html>
  <head>
    <meta name="viewport" content="initial-scale=1.0, user-scalable=no">
    <meta charset="utf-8">
    <title>Traffic Layer</title>
    <style>
      /* Always set the map height explicitly to define the size of the div
       * element that contains the map. */
      #map {
        height: 100%;
      }
      /* Optional: Makes the sample page fill the window. */
      html, body {
        height: 100%;
        margin: 0;
        padding: 0;
      }
    </style>
  </head>
  <body>
    <div id="map"></div>
    <script>

      window.requestAnimationFrame = window.requestAnimationFrame || function () { return true; }

      function initMap() {
          var map = new google.maps.Map(document.getElementById('map'), {
              zoom: 12,
              center: {lat: 40.7128, lng: -74.0060}
          });
          
          var trafficLayer = new google.maps.TrafficLayer();
          trafficLayer.setMap(map);
          
          window.status = 'loaded';
      }

    </script>
    <script async defer
    src="https://maps.googleapis.com/maps/api/js?key=API_KEY&callback=initMap">
    </script>
  </body>
</html>
```

**Comment by exowanderer** [comment]
For better reference, I just tried the exact same code on a local Ubuntu 16.04 machine (with a display) and found the exact same errors (blank PDF or empty white line png).

Here are the system properties in case that's useful:

Distributor ID: Ubuntu
Description: Ubuntu 16.04.5 LTS
Release: 16.04
Codename: Xenial
Architecutre: x86_64

**Comment by leo-moxiworks** [comment]
@exowanderer it seems that your issues are beyond the scope of this ticket and I currently do not have bandwidth to help any further. Maybe you can post your own ticket or see if anyone else is running into the same issues.

**Comment by exowanderer** [comment]
> @exowanderer it seems that your issues are beyond the scope of this ticket and I currently do not have bandwidth to help any further. Maybe you can post your own ticket or see if anyone else is running into the same issues.

@leo-moxiworks: I more than understand that. You gave me a tremendous amount of help about this issue, which _feels_ related to the topic at hand; but may not be.

Hoping not to ask too much: could you share with us here a stripped down javascript snippet, which calls Google Maps in a way that works (for you) with wkhtmltopdf v0.12.5 -- given the additions that you suggested above?

I suspect that you have a code that works, which is where you have been deriving all of your advice. Even though they didn't work; I am still grateful for the effort. If you could share that snippet here (in a full working fashion) that I could confirm if it works for me. I suspect that it will not; but that is the test.

Thank you!

**Comment by leo-moxiworks** [comment]
> > @exowanderer it seems that your issues are beyond the scope of this ticket and I currently do not have bandwidth to help any further. Maybe you can post your own ticket or see if anyone else is running into the same issues.
> 
> @leo-moxiworks: I more than understand that. You gave me a tremendous amount of help about this issue, which _feels_ related to the topic at hand; but may not be.
> 
> Hoping not to ask too much: could you share with us here a stripped down javascript snippet, which calls Google Maps in a way that works (for you) with wkhtmltopdf v0.12.5 -- given the additions that you suggested above?
> 
> I suspect that you have a code that works, which is where you have been deriving all of your advice. Even though they didn't work; I am still grateful for the effort. If you could share that snippet here (in a full working fashion) that I could confirm if it works for me. I suspect that it will not; but that is the test.
> 
> Thank you!

@exowanderer the main difference between our implementation and yours is that we are using the staticmap api and not the javascript api. This might be a requirement for maps to work with wkhtmltopdf. Can you try this instead?

https://maps.googleapis.com/maps/api/staticmap

**Comment by Georgevanvelzen** [comment]
> After banging my head on this for a week, I was able to find this solution that works with wkhtmltopdf version 12.x. Add this code to your javascript to create a requestAnimationFrame function because the old version of QT web does not support this function.
> 
> ```
> window.requestAnimationFrame = window.requestAnimationFrame || function () { return true; }
> ```
> 
> Google maps API is checking if your browser/window can use the requestAnimationFrame method and throwing an error if it cannot. You can override the method and just return true if your webpage does not use the requestAnimationFrame method. This works with google maps API version 3.35+.
> 
> ```
>  if (!_.y.requestAnimationFrame)
>             throw _.U("controls").then(function(b) {
>                 b.pg(a)
>             }),
>             Error("The Google Maps JavaScript API does not support this browser.");
> ```

Hi - thanks for posting this. 

For the past several months I was just using API=xx as part of the command, but this seems to have stopped working this week.

I tried putting your script at the beginning of my code and volia, it started working again. So thanks for saving me a heap of time.

your code (thanks):
window.requestAnimationFrame = window.requestAnimationFrame || function () { return true; }


I did try the WKHTMLTOPDF Bleeding Edge, which did work for generating the Google Map, but it doesn't have the patched QT version, so would not display background colours, etc.

**Comment by gregallensworth** [comment]
Thank you @Georgevanvelzen for your detective work in noting that cause, and an elegant and effective solution.
